Introduce basic support for custom executors

rjmccall opened the pull request:
- Introduce an `UnownedSerialExecutor` type into the concurrency library.
- Create a `SerialExecutor` protocol which allows an executor type to change how it executes jobs.
- Add an `unownedExecutor` requirement to the `Actor` protocol.
- Change the ABI for `Builtin.Executor` (aka `ExecutorRef`, aka `UnownedSerialExecutor`) so that it stores a `SerialExecutor` witness table pointer in the implementation field.  This effectively makes the type an `unowned(unsafe) SerialExecutor`, except that default actors are represented without a witness table pointer.
- Synthesize the `unownedExecutor` method for default actors (i.e. actors that don't provide any other execution logic).
- Make synthesized unownedExecutor properties `@_resilientFinal`, and give them a semantics attribute specifying that they're for default actors.
- Split `Builtin.buildSerialExecutorRef` into a few more precise builtins.  We're not using the main-actor one yet, though.

Pitch thread: https://forums.swift.org/t/support-custom-executors-in-swift-concurrency/44425

rjmccall committed:
Introduce basic support for custom executors.

- Introduce an UnownedSerialExecutor type into the concurrency library.
- Create a SerialExecutor protocol which allows an executor type to
  change how it executes jobs.
- Add an unownedExecutor requirement to the Actor protocol.
- Change the ABI for ExecutorRef so that it stores a SerialExecutor
  witness table pointer in the implementation field.  This effectively
  makes ExecutorRef an `unowned(unsafe) SerialExecutor`, except that
  default actors are represented without a witness table pointer (just
  a bit-pattern).
- Synthesize the unownedExecutor method for default actors (i.e. actors
  that don't provide an unownedExecutor property).
- Make synthesized unownedExecutor properties `final`, and give them
  a semantics attribute specifying that they're for default actors.
- Split `Builtin.buildSerialExecutorRef` into a few more precise
  builtins.  We're not using the main-actor one yet, though.

Pitch thread:
  https://forums.swift.org/t/support-custom-executors-in-swift-concurrency/44425

rjmccall commented:
@swift-ci Please test

ktoso commented on stdlib/public/Concurrency/Executor.swift:
```diff
@@ -12,5 +12,59 @@
+  }
+
+  @inlinable
+  public init<E: SerialExecutor>(ordinary executor: __shared E) {
```
oh, this one I don't know, mind if I ask/learn about the implications of __shared? Is it a borrow?

rjmccall replied:
Yes, it's a borrow.  It's the default argument convention except on initializers.

ktoso replied:
I see, thanks!

ktoso commented on stdlib/public/Concurrency/Actor.swift:
```diff
@@ -19,6 +19,20 @@ import Swift
+  /// eliminated, and rearranged with other work, and they may even
+  /// be introduced when not strictly required.  Visible side effects
+  /// are therefore strongly discouraged within this property.
+  var unownedExecutor: UnownedSerialExecutor { get }
```
yay, happy to see the Actor protocol put to good use with those :) 

And we're following the same pattern in distributed :)

// Complete nitpick for future: I was very very tempted to make those properties be $executor, $transport, $address... to not conflict with user defined things... Would it be weird to define those with `$`? They're _kind of_ synthesized after all... (Allowing this would need to make parsing more permissive, perhaps for some internal marked known names etc hm...)

rjmccall replied:
I deliberately didn't call it `executor` because I wanted to leave space for an owned property (presumably implemented in terms of this one).  I'm not worried about actor types wanting to name something `unownedExecutor`.

ktoso replied:
I see; what would the owned version be preferable for? (I may be missing things to understand the difference here, sorry)

ktoso commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -6211,6 +6222,9 @@ ValueDecl *TypeChecker::deriveProtocolRequirement(DeclContext *DC,
 
+  case KnownDerivableProtocolKind::Actor:
+    return derived.deriveActor(Requirement);
+
```
👍 Distributed does the same... :)

ktoso reviewed (commented):
I'm in no position to provide useful feedback on impl I guess, but read through and at least much more of it nowadays I understood :)

Some minor questions in line... only if you have the time, nothing urgent. Looking great overall!

ktoso commented on lib/SILOptimizer/Mandatory/LowerHopToActor.cpp:
```diff
@@ -106,23 +106,72 @@ bool LowerHopToActor::processHop(HopToExecutorInst *hop) {
-  // TODO: use the Executor protocol
-  auto unmarkedExecutor =
-    B.createBuiltin(loc, builtinName, resultType, subs, {actor});
+  // If the actor type is a default actor, go ahead and devirtualize here.
```
That's pretty cool and in practice the huge majority of actors right? Except obj-c compatible ones and if someone overrides e.g. the executor since that'll trigger `hasExplicitCustomActorMethods` right?

rjmccall replied:
We expect so, yes, with the major exception of the main actor.

rjmccall committed:
Add a utility to clone an `@available` attribute.

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall committed:
Remove the implicit nil inhabitant of Builtin.Executor,
and traffic in Optional<Builtin.Executor> in various places.

jckarter commented on lib/SILGen/SILGenType.cpp:
```diff
@@ -59,8 +59,13 @@ SILGenModule::emitVTableMethod(ClassDecl *theClass,
   if (baseClass == theClass) {
-    // This is a vtable entry for a method of the immediate class.
-    implKind = SILVTable::Entry::Kind::Normal;
+    if (derivedDecl->isResilientFinal())
```
Can you reuse `isNonOverridden` here? In the full generality I ended up needing that to be orthogonal to the existing set of kinds.

rjmccall replied:
Can you remember why?  The modeling seems *much* cleaner this way.

jckarter replied:
I don't completely remember the specifics, but the distinction between `Normal` and `Inherited` is significant, and when a method has no overrides but the class does have other subclasses, then it is perfectly valid to have an entry that is both inherited and non-overridden in the subclass.

jckarter commented:
`@_resilientFinal` seems like it's nearly identical in runtime impact to `[nonoverridden]` for pruned vtable entries. It's nice to have an attribute to force a declaration to remain nonoverridden within its compilation unit, but I'm concerned about it having ABI impact making it externally distinguishable from a `public` non-open declaration without overrides, because that will be yet another edge case we need to deal with forever. It should be straightforward to add support for nonoverridden method descriptors using the representation from https://github.com/apple/swift/pull/36897, since you don't need to be concerned about backward deployment here so can rely on a modification to `swift_lookUpClassMethod` to honor a new attribute flag.

rjmccall commented:
@swift-ci Please test

rjmccall commented:
> `@_resilientFinal` seems like it's nearly identical in runtime impact to `[nonoverridden]` for pruned vtable entries. It's nice to have an attribute to force a declaration to remain nonoverridden within its compilation unit, but I'm concerned about it having ABI impact making it externally distinguishable from a `public` non-open declaration without overrides, because that will be yet another edge case we need to deal with forever.

How do you feel this is distinguished?  Just because it (might) forbid `super` dispatch?

jckarter commented:
IIRC method descriptors are also referenced by external subclasses so we can resiliently build their vtable; would that ever be a concern here? To be clear, I don't think the backward deployment support here is particularly onerous; I had planned to address it in the linked PR by linking in a lookup method implementation that just does the "is nonoverridden" check first then forwards to the OS implementation.

rjmccall committed:
Default actors carry a null witness-table pointer in Builtin.Executor.

Previously, they were storing a low-bit flag that indicated that they
were a default actor.  Using an extra inhabitant frees up the low bit
for future use without being conspicuously more expensive to check.

rjmccall commented:
@swift-ci Please test

rjmccall commented:
> IIRC method descriptors are also referenced by external subclasses so we can resiliently build their vtable; would that ever be a concern here? 

I think we only do this with overrides, and of course these aren't overridable.  I don't think you can safely back-deploy an override in any case.  The existing runtime seems to `fatalError` if you have an override descriptor for something that's not within the bounds.

I do see what I think is your point: that we don't semantically prevent `super` dispatch to arbitrary methods that you haven't overridden, and so the optimizer-driven use case still needs to allow such dispatch, and therefore needs to emit a method descriptor just in case.  Method descriptors are quite small in practice, so I guess most of the code-size hit of allowing that would just be the symbol name being present in the export tables, plus the impact on the lookup function.

I wonder if it would be feasible to treat this as a language defect and ban `super` dispatch to something that you wouldn't be allowed to override in your module.

jckarter commented:
> I think we only do this with overrides, and of course these aren't overridable.

But your goal here is to allow this to become overridable in the future, no? Otherwise it would just be `final`. Going from `public` to `open` should already be ABI compatible.

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please test

rjmccall commented:
@swift-ci Please clean test Windows

drodriguez commented:
I had to kill the Windows+VS2017 CI job that was testing this because it was stuck executing the custom_executors.swift test. It has been 6 hours, when normally the builds are around 1 hour. Just FYI. It might probably get stuck again. I might disable the test in Windows+VS2017 during the weekend.

rjmccall commented:
I've also merged this into the concurrency-5.5-abi-2 branch.

rjmccall commented:
> I had to kill the Windows+VS2017 CI job that was testing this because it was stuck executing the custom_executors.swift test. It has been 6 hours, when normally the builds are around 1 hour. Just FYI. It might probably get stuck again. I might disable the test in Windows+VS2017 during the weekend.

Damn.  The Windows bot was broken earlier, and I just had to hope that that wasn't masking a bug, but I guess it didn't work out.  I can't debug the problem, but I'll look the code over to see if I can figure it out by inspection.

rjmccall commented:
Opened #37200 to disable the test on Windows.  IIUC there are several major bugs in the Windows concurrency support that cause this kind of test failure.

drodriguez commented:
It seems to have been working OK for two builds in a row. It is probably flaky, but not that flaky. I haven't look, but I imagine it is not a logic error, but a small race in the Windows implementation that would be hard to figure out without a lot of trying. It is fine disabling it so it is less noise.
